libethereum/BlockChainSync.cpp

```diff
@@ -0,0 +1,781 @@
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with cpp-ethereum.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/** @file BlockChainSync.cpp
+ * @author Gav Wood <i@gavwood.com>
+ * @date 2014
+ */
+
+#include "BlockChainSync.h"
+
+#include <chrono>
+#include <libdevcore/Common.h>
+#include <libp2p/Host.h>
+#include <libp2p/Session.h>
+#include <libethcore/Exceptions.h>
+#include <libethcore/Params.h>
+#include "BlockChain.h"
+#include "BlockQueue.h"
+#include "EthereumPeer.h"
+#include "EthereumHost.h"
+#include "DownloadMan.h"
+
+using namespace std;
+using namespace dev;
+using namespace dev::eth;
+using namespace p2p;
+
+unsigned const c_chainReorgSize = 30000;
+
+BlockChainSync::BlockChainSync(EthereumHost& _host):
+	m_host(_host)
+{
+	m_bqRoomAvailable = host().bq().onRoomAvailable([this]()
+	{
+		RecursiveGuard l(x_sync);
+		continueSync();
+	});
+}
+
+BlockChainSync::~BlockChainSync()
+{
+	abortSync();
+}
+
+DownloadMan const& BlockChainSync::downloadMan() const
+{
+	return host().downloadMan();
+}
+
+DownloadMan& BlockChainSync::downloadMan()
+{
+	return host().downloadMan();
+}
+
+void BlockChainSync::abortSync()
+{
+	host().foreachPeer([this](EthereumPeer* _p) { onPeerAborting(_p); return true; });
+	downloadMan().resetToChain(h256s());
+}
+
+void BlockChainSync::onPeerStatus(EthereumPeer* _peer)
+{
+	RecursiveGuard l(x_sync);
+	DEV_INVARIANT_CHECK;
+	if (_peer->m_genesisHash != host().chain().genesisHash())
+		_peer->disable("Invalid genesis hash");
+	else if (_peer->m_protocolVersion != host().protocolVersion() && _peer->m_protocolVersion != EthereumHost::c_oldProtocolVersion)
+		_peer->disable("Invalid protocol version.");
+	else if (_peer->m_networkId != host().networkId())
+		_peer->disable("Invalid network identifier.");
+	else if (_peer->session()->info().clientVersion.find("/v0.7.0/") != string::npos)
+		_peer->disable("Blacklisted client version.");
+	else if (host().isBanned(_peer->session()->id()))
+		_peer->disable("Peer banned for previous bad behaviour.");
+	else
+	{
+		unsigned estimatedHashes = estimateHashes();
+		_peer->m_expectedHashes = estimatedHashes;
+		onNewPeer(_peer);
+	}
+	DEV_INVARIANT_CHECK;
+}
+
+unsigned BlockChainSync::estimateHashes()
+{
+	BlockInfo block = host().chain().info();
+	time_t lastBlockTime = (block.hash() == host().chain().genesisHash()) ? 1428192000 : (time_t)block.timestamp;
+	time_t now = time(0);
+	unsigned blockCount = c_chainReorgSize;
+	if (lastBlockTime > now)
+		clog(NetWarn) << "Clock skew? Latest block is in the future";
+	else
+		blockCount += (now - lastBlockTime) / (unsigned)c_durationLimit;
+	clog(NetAllDetail) << "Estimated hashes: " << blockCount;
+	return blockCount;
+}
+
+void BlockChainSync::requestBlocks(EthereumPeer* _peer)
+{
+	if (host().bq().knownFull())
+	{
+		clog(NetAllDetail) << "Waiting for block queue before downloading blocks";
+		m_lastActiveState = m_state;
+		pauseSync();
```

Comment:
The event is triggered once queue has enough space for known blocks. There is another limit for unknown blocks and in case it is reached sync restarts: https://github.com/ethereum/cpp-ethereum/pull/2225/files#diff-9f5b8b62d8ddf0597e54e7c46e236653R220
Future blocks are treated as unknown